gltfTutorial/gltfTutorial_006_MaterialsTechniquesProgramsShaders.md

```diff
@@ -0,0 +1,288 @@
+
+
+## Programs and Shaders
+
+The rendering process in OpenGL is described with small shader programs, which are written in [GLSL, the OpenGL shading language](https://www.opengl.org/documentation/glsl/). Such a shader program consists of at least two shaders, namely the vertex shader and the fragment shader. Roughly speaking, the vertex shader defines how the vertices of an object are transformed, preprocessed and projected on the screen, and the fragment shader defines how these projected vertices are used to fill the pixels on the screen to generate the final rendered image. For further details, refer to one of the GLSL tutorials (e.g. at [open.gl](https://open.gl/drawing#Shaders), [webglfundamentals.org](http://webglfundamentals.org/webgl/lessons/webgl-shaders-and-glsl.html) or others)
+
+### Shaders
+
+In the context of glTF, the vertex- and fragment shader are usually contained in dedicated files, which are referred to by the [`shader`](https://github.com/KhronosGroup/glTF/tree/master/specification#reference-shader) objects using a URI:
+
+```javascript
+"shaders": {
+    "exampleVertexShader": {
+        "uri": "vertexShader.glsl",
+        "type": 35633
+    },
+    "exampleFragmentShader": {
+        "uri": "fragmentShader.glsl",
+        "type": 35632
+    }
+}
+```
+
+The shader source code that is referred to by the URI is stored as a plain text, as described in the section about [GLSL shader data in `shaders`](gltfTutorial_002_BasicGltfStructure.md#glsl-shader-data-in-shaders). This source code can be read into a string, and directly be passed to OpenGL, as shown in the implementation notes below. In addition to the URI, the `shader` object also contains the `type` of the shader, which simply is the respective OpenGL constant: Either `35633` (`GL_VERTEX_SHADER`) for vertex shaders, or `35632` (`GL_FRAGMENT_SHADER`) for fragment shaders.
+
+
+#### Attributes and uniforms in shaders
+
+The vertex shader may define different *attributes*. These attributes describe all properties of a single vertex. Recall this image, already shown in the section about [Mesh primitive attributes](gltfTutorial_005_MeshesTexturesImagesSamplers.md#mesh-primitive-attributes):
+
+![meshPrimitive accessors](images/meshPrimitiveAttributes.png)
+
+One can imagine that the vertex shader is executed for each vertex, and receives the attribute values for the respective vertex in the attribute variables. For example, the following vertex shader defines attributes for the position, normal and texture coordinate of each vertex:
+
+```glsl
+attribute vec3 a_position;
+attribute vec3 a_normal;
+attribute vec2 a_texcoord0;
+
+...
+uniform vec3 u_lightDirection;
+uniform mat4 u_modelViewMatrix;
+...
+
+void main(void)
+{
+    vec4 pos = u_modelViewMatrix * vec4(a_position,1);

```

Comment:
I thought that I once encountered compilers complaining about `1.0` being `double`, and demanding `1.0f`, **but** it might well be that I mixed something up here, and this was not in the context of GLSL, but OpenCL kernels... 

However, I assume that `1.0` is a better practice here, and will review the GLSL spec also regarding the implicit `1` that you mentioned.